### PR TITLE
Add Media Usage Tracker and Sync functionality

### PR DESCRIPTION
**Pull request by shreyasikhar**
## Summary

This PR introduces two new systems for tracking media usage across posts:

- **Live tracking** (`Media_Usage_Tracker`) — automatically tracks which media files are used in which posts on every save, building a reverse index (`_godam_usage_post_ids` on attachments) and a forward index (`_godam_tracked_media` on posts). Also notifies GoDAM Central via `log_media_view` / `remove_media_view` API calls when GoDAM media is added to or removed from a post.
- **Backfill** (`Media_Usage_Backfill`) — one-time background sync for existing posts that predate the tracker, with a "Media Usage Sync" tab in the GoDAM Tools page.
- This media tracking in WordPress will be helpful in Media Versions feature when replacing old version with new version across all the posts

### Issue

https://github.com/rtCamp/godam-core/issues/321

## What's tracked

- All Gutenberg blocks (core and GoDAM), Classic Editor content, and any block registering attachment IDs via the `godam_attachment_ids_from_block` filter
- Both WP-native numeric attachment IDs and GoDAM Central string IDs (`_godam_original_id`)
- Post lifecycle: save, update, permanent delete. Trashed posts remain tracked until permanently deleted.

## Live Tracker highlights

- Two-pass content extraction: HTML URL scanning (covers all blocks generically) + block comment JSON parsing (for blocks with stable attribute schemas)
- GoDAM CDN URL pattern detection (`*.gdcdn.us`) for resolving CDN URLs back to WP attachment IDs
- Diff-based updates on save — only changed attachments trigger meta writes and API calls
- Async GoDAM Central API calls via Action Scheduler with dedup guard; synchronous fallback if AS is unavailable
- Per-request in-memory caching for GoDAM ID → WP post ID and URL → attachment ID lookups
- WP filters: `godam_extracted_attachment_ids`, `godam_attachment_ids_from_block`, `godam_attachment_ids_from_block_attrs`

## Backfill highlights

- Processes posts in batches inside a timed loop (50 posts/batch, 20s wall-clock limit per AS action) — significantly reduces AS invocations for large sites vs. one action per batch
- Queries only posts without `_godam_tracked_media` (NOT EXISTS on indexed `meta_key`) — already-processed posts are skipped; safe to restart after failure
- Includes `stop` endpoint and Resume/Stop UI — progress counters are preserved across stop/resume cycles
- Per-action memory cleanup (`$wpdb->flush()` + `wp_cache_flush_runtime()`) to prevent unbounded memory growth across loop iterations
- Post types: all public post types (via `get_post_types`) excluding `attachment`; filterable via `godam_media_backfill_post_types` and `godam_media_backfill_post_statuses`
- Object-cache layer (`wp_cache_set/delete`) on `get_status()` with explicit invalidation on every write

## REST API (admin-only, `godam/v1`)

| Method | Endpoint | Description |
|--------|----------|-------------|
| `POST` | `/media-backfill/start` | Start or resume the backfill |
| `POST` | `/media-backfill/stop` | Stop an in-progress backfill |
| `GET`  | `/media-backfill/status` | Returns `{status, processed, total, percentage}` |

## UI

New **Media Usage Sync** tab in GoDAM Tools (`/wp-admin/admin.php?page=rtgodam_tools`) with live progress bar, Start/Stop/Resume controls, and 4-second polling during active runs.

| Before Sync |
|--------|
| image |

| After Sync |
|--------|
| image | 

## Test plan

- [ ] Publish a new post with GoDAM and native media — verify `_godam_usage_post_ids` is updated on the attachment and `log_media_view` is scheduled in Action Scheduler
- [ ] Remove media from a post and save — verify `_godam_usage_post_ids` is updated and `remove_media_view` is scheduled
- [ ] Permanently delete a post — verify attachment meta is cleaned up
- [ ] Run Media Usage Sync on an existing site, verify all posts are processed and the sync completes
- [ ] Stop the sync mid-way and resume — verify it continues from where it left off
- [ ] Verify stop/start via WP-CLI:
  ```bash
  wp option get godam_media_backfill_status
  wp eval "var_dump(as_has_scheduled_action('godam_media_backfill_batch'));"

### CODE REVIEW

File: inc/classes/class-media-usage-backfill.php [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+class Media_Usage_Backfill {
+
+	use Singleton;
+
+	// ----- Action Scheduler action name -----
+	const AS_BATCH_ACTION = 'godam_media_backfill_batch';
+
+	// ----- Number of posts processed per batch -----
+	const BATCH_SIZE = 50;
+
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply inc/classes/class-media-usage-backfill.php:42]: Updated the doc, 50 is correct

File: inc/classes/class-media-usage-tracker.php [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+		 *
+		 * Use this to add IDs from completely custom extraction logic (e.g. shortcodes,
+		 * custom fields that embed attachment references outside of post_content).
+		 *
+		 * @param int[]  $attachment_ids Resolved WP attachment post IDs.
+		 * @param string $post_content   Raw post content.
+		 */
+		$attachment_ids = (array) apply_filters( 'godam_extracted_attachment_ids', $attachment_ids, $post_content );
+
+		return array_values( array_unique( array_filter( array_map( 'intval', $attachment_ids ) ) ) );
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply inc/classes/class-media-usage-tracker.php:569]: Updating the doc

File: inc/classes/class-media-usage-tracker.php [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+	 *
+	 * @param int $attachment_id Attachment WP post ID.
+	 * @param int $post_id       Post ID using the attachment.
+	 * @return void
+	 */
+	private function add_post_to_attachment( $attachment_id, $post_id ) {
+		$ids   = $this->get_usage_post_ids( $attachment_id );
+		$ids[] = (int) $post_id;
+		update_post_meta( $attachment_id, self::ATTACHMENT_META_KEY, array_values( array_unique( $ids ) ) );
+	}
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply inc/classes/class-media-usage-tracker.php]: Updated the doc

File: inc/classes/class-media-usage-tracker.php [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+	 *
+	 * @param int $attachment_id Attachment WP post ID.
+	 * @param int $post_id       Post ID to remove.
+	 * @return void
+	 */
+	private function remove_post_from_attachment( $attachment_id, $post_id ) {
+		$ids     = $this->get_usage_post_ids( $attachment_id );
+		$new_ids = array_values( array_diff( $ids, array( (int) $post_id ) ) );
+		update_post_meta( $attachment_id, self::ATTACHMENT_META_KEY, $new_ids );
+	}
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply inc/classes/class-media-usage-tracker.php]: sorting is not needed, updated the doc

File: pages/tools/components/tabs/MediaUsageSyncTab.jsx [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+	const startPolling = () => {
+		if ( pollRef.current ) {
+			return;
+		}
+		pollRef.current = setInterval( async () => {
+			const nextStatus = await fetchStatus();
+			if ( nextStatus !== 'running' ) {
+				stopPolling();
+				if ( nextStatus === 'completed' ) {
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply pages/tools/components/tabs/MediaUsageSyncTab.jsx:68]: addressed https://github.com/rtCamp/godam/pull/1919/commits/8011bf2c8cb673f1211f63111f8739cb70ebaf40

File: inc/classes/class-media-usage-backfill.php [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+			return;
+		}
+
+		$total = $this->count_pending_posts();
+
+		update_option( self::OPT_STATUS, self::STATUS_RUNNING, false );
+		update_option( self::OPT_PROCESSED, 0, false );
+		update_option( self::OPT_TOTAL, $total, false );
+		wp_cache_delete( self::CACHE_KEY, self::CACHE_GROUP );
+
```
[the remark this answers is not in the document]
Comment (shreyasikhar) [reply inc/classes/class-media-usage-backfill.php]: addressed: https://github.com/rtCamp/godam/pull/1919/commits/fc670ccd34919a70c2dfec129760c25ecfaaecfd

**Review by elifvish — APPROVED** [review]

**Comment by subodhr258** [comment]
Solid, thorough implementation — the reference-counted source map (so the content scanner and the Woo add-on can share one index) and the timed-loop batching in the backfill are nicely thought through. A few things worth a look before merge:

- **`save_post` runs the full extraction synchronously on every post of every type (priority 99).** The `empty( $post_content )` early return covers most cases, but a non-empty post still pays `parse_blocks()` + the HTML regex passes + an `_elementor_data` meta read on every save — including high-frequency CPT saves like WooCommerce orders. Worth confirming the per-save overhead is acceptable on large/busy sites.

- **`log_media_view` vs `remove_media_view` auth looks inconsistent.** Both bail early when `rtgodam-api-key` is empty, but `async_log_media_view` sends no `X-Api-Key` (comment says "unauthenticated endpoint") while `async_remove_media_view` does send it. If that asymmetry is intentional, a one-line comment on the remove side explaining why would help; otherwise the two should match.

- **No automated tests** for `extract_attachment_ids()` and the resolvers. That parsing layer (HTML regex / block attrs / shortcodes / Elementor / CDN URL → attachment ID) is the riskiest part of the PR and would benefit from a few unit tests around the resolution edge cases.

**Review by subodhr258 — APPROVED** [review]

**Comment by subodhr258** [comment]
Code LGTM, assuming it's self-tested

File: inc/classes/class-media-usage-backfill.php [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+	 *                     switches to it before running, because Action Scheduler
+	 *                     fires jobs in the main-site context regardless of which
+	 *                     subsite queued them. 0/absent on single-site.
+	 * @return void
+	 */
+	public function process_batch( $blog_id = 0 ) {
+		$blog_id  = (int) $blog_id;
+		$switched = false;
+		if ( is_multisite() && $blog_id > 0 && get_current_blog_id() !== $blog_id ) {
+			switch_to_blog( $blog_id );
```
Comment (rtBot) [inline inc/classes/class-media-usage-backfill.php]: :warning: **Warning**: switch_to_blog() may not work as expected since it only changes the database context for the blog and does not load the plugins or theme of that site. Filters or hooks on the blog you are switching to will not run (*WordPressVIPMinimum.Functions.RestrictedFunctions.switch_to_blog_switch_to_blog*).

**Review by rtBot — COMMENTED** [review]
# Code analysis identified issues
[action-phpcs-code-review](https://github.com/rtCamp/action-phpcs-code-review/) has identified potential problems in this pull request during automated scanning. We recommend reviewing the issues noted and that they are resolved.

**phpcs** scanning turned up:

:warning: 1 warning 


***


Powered by rtCamp's [GitHub Actions Library](https://github.com/rtCamp/github-actions-library)

**Review by subodhr258 — APPROVED** [review]
